### PR TITLE
Replace def_id_no_primitives with def_id

### PR DESCRIPTION
**Pull request by zredb**
this is my first PR for this repository, please let me know it has ayn question.
r?

### CODE REVIEW

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 1]
```diff
@@ -47,7 +48,7 @@ crate fn collect_trait_impls(mut krate: Crate, cx: &mut DocContext<'_>) -> Crate
                 inline::build_impl(cx, None, def_id, None, &mut new_items);
 
                 // FIXME(eddyb) is this `doc(hidden)` check needed?
-                if !cx.tcx.is_doc_hidden(def_id) {
+                if !cx.tcx.get_attrs(def_id).lists(sym::doc).has_word(sym::hidden) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/collect_trait_impls.rs:51]: This shouldn't have changed.

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 2]
```diff
@@ -64,6 +65,7 @@ crate fn collect_trait_impls(mut krate: Crate, cx: &mut DocContext<'_>) -> Crate
         map: &FxHashMap<DefId, &Type>,
         cleaner: &mut BadImplStripper,
         type_did: DefId,
+        c: &Cache,
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/collect_trait_impls.rs:68]: `cx` is already passed in the parameters, so no need to add the `cache` in here. Just call `cx.cache`.

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 3]
```diff
@@ -176,7 +185,13 @@ impl<'a, 'tcx> DocVisitor for SyntheticImplCollector<'a, 'tcx> {
     fn visit_item(&mut self, i: &Item) {
         if i.is_struct() || i.is_enum() || i.is_union() {
             // FIXME(eddyb) is this `doc(hidden)` check needed?
-            if !self.cx.tcx.is_doc_hidden(i.def_id.expect_def_id()) {
+            if !self
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/collect_trait_impls.rs:188]: This shouldn't have changed.

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 4]
```diff
@@ -211,13 +226,13 @@ struct BadImplStripper {
 }
 
 impl BadImplStripper {
-    fn keep_impl(&self, ty: &Type, is_deref: bool) -> bool {
+    fn keep_impl(&self, ty: &Type, is_deref: bool, c: &Cache) -> bool {
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/collect_trait_impls.rs:229]: Please name it `cache` and not just `c`.

File: src/librustdoc/clean/types.rs [hunk 5]
```diff
@@ -1140,6 +1140,15 @@ impl PartialEq for Attributes {
 
 impl Eq for Attributes {}
 
+impl Hash for Attributes {
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/types.rs:1143]: Why is this needed?

File: src/librustdoc/clean/types.rs [hunk 6]
```diff
@@ -904,7 +904,7 @@ impl<I: Iterator<Item = ast::NestedMetaItem> + IntoIterator<Item = ast::NestedMe
 /// Included files are kept separate from inline doc comments so that proper line-number
 /// information can be given when a doctest fails. Sugared doc comments and "raw" doc comments are
 /// kept separate because of issue #42760.
-#[derive(Clone, PartialEq, Eq, Debug)]
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/types.rs:907]: Why is this needed?

**Comment by GuillaumeGomez** [comment]
Looks good overall. Just a few improvements and it'll be good to go!

**Comment by camelid** [comment]
Please squash your commits too using `git reset --soft HEAD~ && git commit --amend --no-edit` :)

**Comment by camelid** [comment]
You'll also need to [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts) so this PR can be merged.
